File: cplus/leetcode/202510/algo1610.cpp
#include <bits/stdc++.h>
using namespace std;

// 滑动窗口与双指针 - 不定长滑动窗口 2.1.2 进阶（选做）
class Solution {
public:
    int visiblePoints(vector<vector<int>>& points, int angle, vector<int>& location) {
        // 学习极角计算
        auto cal_degree = [](int x, int y) -> double {
            double rad = atan2(y, x);           // (-π, π]
            double deg = rad * 180.0 / M_PI;    // (-180°, 180°]
            if (deg < 0)  deg += 360.0;
            return deg;                         // [0, 360)
        };

        auto x0 = location[0], y0 = location[1];
        int locs = 0;
        vector<double> deg_list;
        for (auto& p: points) {
            if (x0 == p[0] && y0 == p[1]) locs ++;
            else deg_list.push_back(cal_degree(p[0]-x0, p[1]-y0));
        }
        
        int n = deg_list.size(), l = 0, ans = 0;
        sort(deg_list.begin(), deg_list.end());
        for (int i=0; i<n; ++i) deg_list.push_back(deg_list[i]+360);
        
        for (int r=0; r<deg_list.size(); ++r) {
            while (deg_list[l] + angle < deg_list[r]) l ++;
            ans = max(ans, r - l + 1);
        }
        return ans + locs;

    }
};

File: python/algo/202510/algo1610.py
from math import atan2, degrees
from typing import List

# 滑动窗口与双指针 - 不定长滑动窗口 2.1.2 进阶（选做）
class Solution:
    def visiblePoints(self, points: List[List[int]], angle: int, location: List[int]) -> int:
        # 计算极角的度数, 浮点数 [0,360)
        def cal_degree(x:int, y:int) -> float:
            rad = atan2(y, x)
            deg = degrees(rad)
            if deg < 0:
                deg += 360
            return deg  # 结果为0°到359.999...°
    
        x0, y0 = location
        deg_list, locs = [], 0
        for x, y in points:
            if x == x0 and y == y0: locs += 1   # 极点一定包含
            else:deg_list.append(cal_degree(x-x0,y-y0))
        deg_list.sort()
        deg_list += [x+360 for x in deg_list] # 循环列表

        m = len(points)
        l = ans = 0
        for r, d in enumerate(deg_list):
            while deg_list[l] + angle < d:
                l += 1
            ans = max(ans, r-l+1)
            if (ans + locs) == m:return m
        return ans + locs